File: current_session/1143.py
"""
1143
"""
class Solution:
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:
        n,m=len(text1),len(text2)
        dp=[[0]*(m+1) for _ in range(n+1)]
        for i in range(n):
            for j in range(m):
                if text1[i]==text2[j]:
                    dp[i+1][j+1]=dp[i][j]+1
                else:
                    dp[i+1][j+1]=max(dp[i][j+1],dp[i+1][j])
        return dp[-1][-1]

print(Solution().longestCommonSubsequence('abcde','ace'),'is 3')
print(Solution().longestCommonSubsequence('abc','abc'),'is 3')
print(Solution().longestCommonSubsequence('abc','def'),'is 0')
print(Solution().longestCommonSubsequence("ezupkr","ubmrapg"),'is 2')
print(Solution().longestCommonSubsequence("abcba","abcbcba"),'is 5')
print(Solution().longestCommonSubsequence("oxcpqrsvwf","shmtulqrypypppp"),'is 2')
print(Solution().longestCommonSubsequence("abcba","abcbcba"), 'is 5')
print(Solution().longestCommonSubsequence("kvwrkharmnqpwxyhejgvybifmncdorglsfqlidupyvcnypwvglormj","irmdqnwnelyturkdobypezwvonqpubedetansrkjgzyzgpuxajgdaji"),'is 15')
print(Solution().longestCommonSubsequence("lmrejgzsbqpkdonytkbknstsxifofmrktcpq","hklcrebcjipetgnmlqvijovmlgripwratarmt"),'is 12')
print(Solution().longestCommonSubsequence("fcvqfcnglshwpgxujwrylqzejmdubkxs","bctsfwdelkdqzshupmrufyxklsjurevip"),'is 11')
print(Solution().longestCommonSubsequence("tufgfqlspqpidwrmjexifslkzobcjqvwlevghglynojchkvufqnwxixqnercbabm","xuhadmbsbabqzirgrcxazcxypmjebgxyzmlepcdezwbsjkjalgdotcjavojehsvax"),'is 20')

File: current_session/c++/1143.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  int longestCommonSubsequence(string text1, string text2) {
    // use DP
    // dp[i][j] = ans of text1[:i] and text2[:j]
    // if text1[i] == text2[j]: dp[i][j] = dp[i-1][j-1] + 1
    int M = text1.size();
    int N = text2.size();
    vector<vector<int>> dp(M+1, vector<int>(N+1, 0));
    for (int i=0; i<M; ++i) {
      for (int j=0; j<N; ++j) {
        if (text1[i] == text2[j]) {
          dp[i+1][j+1] = dp[i][j]+1;
        } else {
          dp[i+1][j+1] = max(dp[i][j+1], dp[i+1][j]);
        }
      }
    }
    return dp[M][N];
  }
};

int main() {
  cout << Solution().longestCommonSubsequence("abcde", "ace") << endl;
}